feat(platform-connectors): add node metadata enrichment for health ev…

Gyan172004 opened the pull request:
## Summary

**Implements Feature Request #119**

This PR enriches health events with node metadata (provider IDs and labels) before sending them to connectors. This makes it easier to track and correlate GPU health issues across multiple clusters, especially when node names aren't unique.

**Key additions:**
- Fetches node metadata from Kubernetes API (`CoreV1().Nodes().Get()`) with configurable timeout and rate limiting
- Decodes cloud provider IDs into structured fields (extracts region, zone, instance ID from provider-specific formats)
- Implements LRU cache with TTL using `hashicorp/golang-lru` to minimize API load (default: 50 entries, 1-hour expiration)
- Adds double-check locking pattern with mutex to prevent duplicate API calls when multiple events arrive for the same node simultaneously
- Integrates at the platform connector level - enriches events before fanout to MongoDB, Kubernetes connector, and other backends
- Configurable allowlist of 17 default node labels covering topology, instance types, GPU metadata, and physical infrastructure
- Graceful error handling - metadata fetch failures don't block health event processing (logged as warnings)


## Type of Change
- [ ] 🐛 Bug fix
- [X] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Health Monitors
- [X] Core Services
- [ ] Fault Management
- [ ] Documentation/CI
- [ ] Other: ____________

## Testing
- [X] Tests pass locally
- [ ] Manual testing completed
- [X] No breaking changes (or documented)

## Checklist
- [X] Self-review completed
- [X] Documentation updated (if needed)
- [X] Ready for review


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added node metadata enrichment to health events, automatically capturing cloud provider IDs, topology labels, instance types, CUDA driver/runtime versions, and custom Kubernetes labels across AWS, GCP, Azure, and OCI environments.

* **Documentation**
  * Updated configuration documentation with node metadata enrichment settings, including cache configuration and allowed labels.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Gyan172004 committed:
feat(platform-connectors): add node metadata enrichment for health events

Enriches health events with provider IDs and node labels before sending
to connectors. This helps track and correlate events across clusters.

- Add provider ID decoding for AWS, GCP, Azure, OCI
- Include 17 default node labels (topology, instance type, GPU info)
- Use hashicorp/golang-lru for caching (size: 50, TTL: 1 hour)
- Simple mutex with double-check pattern to prevent duplicate API calls
- Best-effort approach - failures don't block event processing
- 37 unit tests with 91.7% coverage, race-free

Fixes #119

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

lalitadithya commented on platform-connectors/pkg/nodemetadata/README.md:
Let's not maintain package level readmes, instead let's try and make the code as self documenting as possible

Gyan172004 replied:
Removed the package-level README. The code is now self-documenting and updated in main README.

lalitadithya commented on platform-connectors/pkg/nodemetadata/cache.go:
```diff
@@ -0,0 +1,97 @@
+		return metadata, nil
+	}
+
+	// Slow path: acquire lock and check again
```
Do we need these obvious comments? Isn't the code self-explanatory in most cases?

lalitadithya commented on platform-connectors/pkg/nodemetadata/cache_test.go:
```diff
@@ -0,0 +1,490 @@
+	}
+}
+
+// ============================================================================
```
These tests confuse me; these seem to be testing the functionality of the OSS package which is already tests. Can you please explain a bit further on why we need these tests?

Gyan172004 replied:
Understood, there is no need to test the already-tested OSS package, deleted cache_test.go.

lalitadithya commented on platform-connectors/pkg/nodemetadata/config.go:
```diff
@@ -0,0 +1,168 @@
+	"nvidia.com/gpu.deploy.driver",       // Driver deployment topology
+	"nvidia.com/gpu.count",               // Number of GPUs
+	"nvsentinel.dgxc.nvidia.com/dcgm.version",     // DCGM version tracking
+	"nvsentinel.dgxc.nvidia.com/driver.installed", // Driver installation status
```
We don't need these, can we instead pick the ones that have cuda/driver versions on them?

lalitadithya replied:
Can you check the official GPU operator docs on their github repo and confirm the labels, please?

lalitadithya commented on platform-connectors/pkg/nodemetadata/processor.go:
```diff
@@ -0,0 +1,153 @@
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+
+	restConfig, err := rest.InClusterConfig()
```
We need to build from flags using clientcmd. I suspect we already have another client that is already built with the correct QPS etc, let's use the same.

lalitadithya commented on platform-connectors/pkg/nodemetadata/processor_test.go:
```diff
@@ -0,0 +1,697 @@
+		},
+	}
+
+	clientset := fake.NewSimpleClientset(node)
```
no -- use envtest like it is done in the other packages

Gyan172004 replied:
Okay, so I should use envtest for integration tests. Should I remove these unit tests using fake nodes?

Gyan172004 replied:
Converted all tests to use envtest with testmain setup. Removed all fake.NewSimpleClientset usage.

Gyan172004 committed:
Address PR feedback - simplify and clean up implementation

- Remove cache abstraction layer (use LRU directly in processor)
- Remove provider ID decoding (store raw IDs only)
- Reuse shared K8s client instead of creating new one
- Remove deprecated beta Kubernetes labels
- Convert all tests to use envtest (no fake clients)
- Clean up self-explanatory comments
- Move default labels to Helm chart
- Remove package-level README (code is self-documenting)

All changes tested and verified with 89.4% code coverage.

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 commented on platform-connectors/pkg/nodemetadata/processor.go:
```diff
@@ -0,0 +1,153 @@
+
+	for _, labelKey := range p.config.AllowedLabels {
+		if labelValue, exists := metadata.Labels[labelKey]; exists {
+			event.Metadata[fmt.Sprintf("node.label.%s", labelKey)] = labelValue
```
Removed, Labels are now stored directly as Kubernetes provides them (e.g., "topology.kubernetes.io/zone").

Gyan172004 committed:
docs: update label info in metadata section

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

lalitadithya commented on platform-connectors/pkg/nodemetadata/nodemetadata.go:
Let's call this `platform-connectors/pkg/nodemetadata/types.go`

Gyan172004 replied:
done

lalitadithya commented on platform-connectors/pkg/nodemetadata/processor_test.go:
```diff
@@ -0,0 +1,671 @@
+	}
+}
+
+func TestProcessorCachingBehavior(t *testing.T) {
```
I don't follow this test, what is this supposed to be testing?

lalitadithya commented on platform-connectors/pkg/nodemetadata/processor.go:
since this processor is specific to kube, let's call it 

platform-connectors/pkg/nodemetadata/kube.go

Gyan172004 replied:
done

lalitadithya commented:
/ok to test f5c549883cc36b5be9d00158dc16bd02ae0bcf22

XRFXLP commented:
/ok to test 3ba27ad50162e88f113189430242f86de2ee6cac

Gyan172004 committed:
feat: Add node metadata enrichment with all review feedback

- Implement node metadata augmentation for health events
- Add factory pattern for processor creation
- Rename processor.go -> kube.go, nodemetadata.go -> types.go
- Add comprehensive table-driven tests with testify/assert
- Fix MongoDB BSON serialization for protobuf structs
- Update README metadata description
- Add CUDA driver/runtime version labels to allowedLabels
- Graceful degradation when K8s connector disabled
- All PR review feedback addressed

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 commented on platform-connectors/pkg/nodemetadata/processor.go:
```diff
@@ -0,0 +1,157 @@
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+
+	if clientset == nil {
```
yes, done

Gyan172004 commented on platform-connectors/pkg/nodemetadata/processor.go:
```diff
@@ -0,0 +1,157 @@
+	}
+
+	if metadata.ProviderID != "" {
+		event.Metadata["node.providerID"] = metadata.ProviderID
```
corrected

Gyan172004 commented on platform-connectors/pkg/nodemetadata/processor_test.go:
```diff
@@ -0,0 +1,671 @@
+}
+
+// cleanupTestNode deletes a node from the test API server
+func cleanupTestNode(t *testing.T, nodeName string) {
```
done

Gyan172004 commented on distros/kubernetes/nvsentinel/values.yaml:
```diff
@@ -108,5 +108,52 @@ platformConnector:
+
+# Affinity rules for platform connector pods
+affinity: {}
+
```
idk it came while resolving merge conflicts.
removed these changes

lalitadithya commented on platform-connectors/pkg/nodemetadata/kube.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+// LRU library handles TTL expiration automatically.
+func (p *processor) Start(ctx context.Context) {
```
It's overkill, let's remove it.

lalitadithya commented:
DCO is failing, can you please check?

lalitadithya commented on data-models/pkg/model/health_event_extentions.go:
```diff
@@ -53,3 +55,144 @@ type HealthEventWithStatus struct {
+
+// MarshalBSON implements custom BSON marshaling for HealthEventWithStatus.
+// This is necessary because protobuf-generated structs don't have BSON tags,
+// so we need to manually convert the HealthEvent to a BSON-compatible format.
```
I don't follow this. So far on main we've never had this issue that needs manual marshalling/unmarshalling. The library was smart enough to skip the non-exported. Can you please elaborate on what changed recently that now requires us to make this change?

Gyan172004 replied:
Its working withour marshalling as well now :)
Removed that part

lalitadithya commented on distros/kubernetes/nvsentinel/values-tilt.yaml:
```diff
@@ -15,6 +15,7 @@
 global:
   dryRun: false
   kubeVersion: 1.31.0
+  installMongoDB: true
```
This is not needed with the recent changes on main. This is likely a remanent of a bad merge, can you please check?

lalitadithya commented on kubelogin-darwin-arm64.zip:
This should not be checked in, plesae remove

lalitadithya commented on platform-connectors/pkg/nodemetadata/factory.go:
```diff
@@ -0,0 +1,91 @@
+)
+
+// ProcessorCreator is a function that creates a processor of a specific type.
+type ProcessorCreator func(ctx context.Context, config *Config, params interface{}) (Processor, error)
```
I don't follow this, we know what types of processors we have in code, why can't we call the initialize functions when required? Do we need to make it this generic?

lalitadithya commented on platform-connectors/pkg/nodemetadata/kube.go:
```diff
@@ -0,0 +1,190 @@
+// newKubernetesProcessor creates a new Kubernetes-based node metadata processor.
+// This function is package-private and called by the factory.
+func newKubernetesProcessor(ctx context.Context, config *Config, params interface{}) (Processor, error) {
+	if config == nil {
```
When would this be nil?

lalitadithya replied:
bump

lalitadithya commented on platform-connectors/pkg/nodemetadata/kube_test.go:
```diff
@@ -0,0 +1,487 @@
+	}
+	require.NoError(t, p.AugmentHealthEvent(ctx, event1))
+
+	// Second call - cache hit
```
How do we know it was picked up from cache?

lalitadithya commented on tilt/Tiltfile:
Why was this deleted?

Gyan172004 committed:
Address review comments: simplify code and remove custom BSON marshaling

- Remove custom MarshalBSON/UnmarshalBSON (default BSON marshaler works)
- Remove over-engineered factory pattern (factory.go, factory_test.go)
- Remove unnecessary Start/Stop lifecycle methods
- Revert MongoDB driver to v1.17.4 (consistent with main)
- Remove dead code (clientset nil check, rebuild comment)
- Remove obvious comments from values files
- Remove emoji from logs
- Improve test assertions and comments
- Delete kubelogin binary and update .gitignore
- Restore accidentally deleted Tiltfile

All tests pass and end-to-end verification confirms metadata enrichment works.

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Remove obvious comments that repeat code intent

Keep only comments for complex logic.

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Update values.yaml

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Update values.yaml

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Final Update values.yaml

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Final Update values.yaml

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Delete data-models/pkg/model/health_event_extentions_test.go

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Clean up test comments to focus on intent

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Add graceful degradation when K8s connector unavailable

When node metadata enrichment is enabled but the Kubernetes clientset
is unavailable, log a warning and continue service startup without
node metadata enrichment instead of aborting.

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Address review feedback: fix unreachable code and improve consistency

- Remove unreachable nil clientset check (service aborts earlier if K8s fails)
- Make node metadata processor truly optional (log warning instead of returning error)
- Use consistent terminology (enrichment vs augmentation)
- Add test for multiple nodes enrichment scenario
- Add comments explaining architectural decisions
- Explicit nil assignment for clarity when processor init fails

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Remove obvious comments from config example

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Simplify factory pattern per review feedback

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Trigger DCO check

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

lalitadithya commented on platform-connectors/pkg/nodemetadata/kube.go:
```diff
@@ -0,0 +1,152 @@
+		return nil, fmt.Errorf("config cannot be nil")
+	}
+
+	if clientset == nil {
```
Same question as earlier

Gyan172004 replied:
Coderabbit mentioned to handle this everywhere.
I will remove it, clientset will always be non nil

Gyan172004 committed:
Clean up merge conflict markers

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Address review feedback: simplify factory, remove unnecessary dependencies

- Remove kubelogin*.zip from .gitignore (not project-specific)
- Remove mongo-driver dependency from data-models (only BSON tags needed)
- Implement platform-based factory pattern with switch statement (like CSP client)
- Remove unreachable nil clientset check from newKubernetesProcessor
- Update tests to match new factory signature and remove nil clientset cases
- Add test for unsupported platform scenario

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 commented on platform-connectors/pkg/nodemetadata/factory.go:
```diff
@@ -0,0 +1,27 @@
+
+// NewProcessor creates a node metadata processor.
+// Currently supports Kubernetes; can be extended for other platforms (e.g., Slurm) in the future.
+func NewProcessor(ctx context.Context, config *Config, clientset kubernetes.Interface) (Processor, error) {
```
done, implementef factory design in switch style

lalitadithya commented on distros/kubernetes/nvsentinel/values-tilt.yaml:
```diff
@@ -133,3 +133,10 @@ csp-health-monitor:
+# Enable node metadata enrichment for testing
+platformConnector:
+  mongodbStore:
+    enabled: true
```
mongodb is already enabled with the global flag now, please check on main. We only need to enable the nodeMetadata

Gyan172004 replied:
Removed. Now only nodeMetadata.enabled under platformConnector.

lalitadithya commented on platform-connectors/main.go:
```diff
@@ -154,35 +192,37 @@ func initializeConnectors(
 	lis net.Listener,
 	k8sRingBuffer *ringbuffer.RingBuffer,
 	storeConnector *store.MongoDbStoreConnector,
+	processor nodemetadata.Processor,
```
I don't think this arg is used in this function, can you please check and remove if it is not used?

lalitadithya reviewed (commented):
looks much better now, please rebase.

Gyan172004 commented on platform-connectors/pkg/nodemetadata/kube_test.go:
```diff
@@ -0,0 +1,471 @@
+	}
+	require.NoError(t, p.AugmentHealthEvent(ctx, event1))
+
+	// Second call for same node - verify cache returns consistent results
```
This was just to check consistency.
I will add to test cache as well, delete previous result and verify no more api call is made.

Gyan172004 committed:
Address final review feedback

- Remove redundant mongodbStore.enabled from platformConnector section in values-tilt.yaml
- Remove explicit processor = nil assignment in initializeK8sConnector (already nil on error)
- Remove unused processor parameter from cleanupResources function
- Remove unnecessary nil config check from newKubernetesProcessor (caller guarantees non-nil)
- Update cache test to delete node after first call, proving cache usage
- Merge TestProcessorMultipleNodesEnrichment into table-driven test structure

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 commented on platform-connectors/main.go:
```diff
@@ -76,34 +78,43 @@ func loadConfig(configFilePath string) (map[string]interface{}, error) {
+	processor, err := initializeNodeMetadataProcessor(ctx, config, clientset)
+	if err != nil {
+		slog.Warn("Failed to initialize node metadata processor, continuing without enrichment", "error", err)
+		processor = nil
```
removed the redundant assignment.

lalitadithya commented:
@Gyan172004 , can you please include https://github.com/NVIDIA/NVSentinel/issues/119#issuecomment-3484128843 in this PR?

Gyan172004 committed:
Add cloud-specific topology labels (AWS/OCI/GCP) to allowlist

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Merge branch 'NVIDIA:main' into feat/node-metadata-enrichment

lalitadithya commented:
/ok to test 5fd4dbaa4e28adcdb2e9786b6cae13745f01480f

Gyan172004 committed:
Fix linting errors: add blank lines and format imports

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

lalitadithya committed:
Merge branch 'main' into feat/node-metadata-enrichment
